packages/amplify-cli/src/execution-manager.ts

```diff
@@ -140,64 +141,94 @@ async function selectPluginForExecution(context: Context, pluginCandidates: Plug
-    ) {
-      attachContextExtensions(context, plugin);
-      await pluginModule.executeAmplifyCommand(context);
-    } else {
-      // if the module does not have the executeAmplifyCommand method,
-      // fall back to the old approach by scanning the command folder and locate the command file
-      let commandFilepath = path.normalize(path.join(plugin.packageLocation, 'commands', plugin.manifest.name, context.input.command!));
-      if (context.input.subCommands && context.input.subCommands.length > 0) {
-        commandFilepath = path.join(commandFilepath, ...context.input.subCommands!);
-      }
+  let commandName = constants.ExecuteAmplifyCommand;
+  let fallbackFn = legacyCommandExecutor;
 
-      let commandModule;
+  if (isHeadlessCommand(context)) {
+    commandName = constants.ExecuteAmplifyHeadlessCommand;
+    fallbackFn = () => context.print.error(`Headless mode is not implemented for ${plugin.packageName}`);
+  }
 
-      try {
-        commandModule = require(commandFilepath);
-      } catch (e) {
-        // do nothing
-      }
+  return executionHandler(context, plugin)
+    .forCommand(commandName)
+    .withFallback(fallbackFn)
+    .run();
+}
 
-      if (!commandModule) {
-        commandFilepath = path.normalize(path.join(plugin.packageLocation, 'commands', plugin.manifest.name));
-        try {
-          commandModule = require(commandFilepath);
-        } catch (e) {
-          // do nothing
+const executionHandler = (context: Context, plugin: PluginInfo) => ({
+  forCommand: (commandName: string) => ({
+    withFallback: (fallbackFn: (context: Context, plugin: PluginInfo) => Promise<void>) => ({
+      run: async (): Promise<void> => {
+        const handler = getHandler(require(plugin.packageLocation), commandName);
+        await raisePreEvent(context);
+        if (!handler) {
+          await fallbackFn(context, plugin);
+        } else {
+          attachContextExtensions(context, plugin);
+          await handler(context);
         }
-      }
+        await raisePostEvent(context);
+      },
+    }),
+  }),
+});
+
+const getHandler = (pluginModule: any, handlerName: string): ((context: Context) => Promise<void>) | undefined => {
+  if (pluginModule.hasOwnProperty(handlerName) && typeof pluginModule[handlerName] === 'function') {
+    if (handlerName === constants.ExecuteAmplifyHeadlessCommand) {
+      return async context => pluginModule[handlerName](context, await readHeadlessPayload());
+    } else {
+      return pluginModule[handlerName];
+    }
+  }
+};
 
-      if (commandModule) {
-        attachContextExtensions(context, plugin);
-        await commandModule.run(context);
-      } else {
-        const { showAllHelp } = require('./extensions/amplify-helpers/show-all-help');
-        showAllHelp(context);
-      }
+// old plugin execution approach of scanning the command folder and locating the command file
+const legacyCommandExecutor = async (context: Context, plugin: PluginInfo) => {
```

Comment:
For what cases do we still need this legacyCommandExecutor function? From what I see all the use-cases should be covered in the new "executionHandler"  function you created above..or am I missing something?

Reply:
AFAIK it's not used, but Weimin left it in when doing the plugin refactor so maybe it's still used by something. I figured it was safest to leave it in unless we're sure it's not used anymore

Reply:
Okay, let's make a note stating that's not in use or something and we can remove it as a part of a separate PR. I wouldn't just have not-used code hanging in there.

Reply:
left a TODO comment